pubsub.go

```diff
@@ -284,31 +361,121 @@ func (p *PubsubValueStore) Cancel(name string) (bool, error) {
 		delete(p.subs, name)
 	}
 
+	if cancel, ok := p.cancels[name]; ok {
+		cancel()
+		delete(p.cancels, name)
+	}
+
 	return ok, nil
 }
 
-func (p *PubsubValueStore) handleSubscription(sub *pubsub.Subscription, key string, cancel func()) {
+func (p *PubsubValueStore) handleSubscription(ctx context.Context, sub *pubsub.Subscription, key string, cancel func()) {
 	defer sub.Cancel()
 	defer cancel()
 
+	newMsg := make(chan []byte)
+	go func() {
+		defer close(newMsg)
+		for {
+			data, err := p.handleNewMsgs(ctx, sub, key)
+			if err != nil {
+				return
+			}
+			select {
+			case newMsg <- data:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	newPeerData := make(chan []byte)
+	go func() {
+		defer close(newPeerData)
+		for {
+			data, err := p.handleNewPeer(ctx, sub, key)
+			if err == nil {
+				if data != nil {
+					select {
+					case newPeerData <- data:
+					case <-ctx.Done():
+						return
+					}
+				}
+			} else {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					log.Errorf("PubsubPeerJoin: error interacting with new peer", err)
+				}
+			}
+		}
+	}()
+
 	for {
-		msg, err := sub.Next(p.ctx)
-		if err != nil {
-			if err != context.Canceled {
-				log.Warningf("PubsubResolve: subscription error in %s: %s", key, err.Error())
+		var data []byte
+		var ok bool
+		select {
+		case data, ok = <-newMsg:
+			if !ok {
+				return
+			}
+		case data, ok = <-newPeerData:
+			if !ok {
+				return
 			}
+		case <-ctx.Done():
 			return
 		}
-		if p.isBetter(key, msg.GetData()) {
-			err := p.ds.Put(dshelp.NewKeyFromBinary([]byte(key)), msg.GetData())
+
+		if p.compare(key, data) > 0 {
+			err := p.ds.Put(dshelp.NewKeyFromBinary([]byte(key)), data)
 			if err != nil {
 				log.Warningf("PubsubResolve: error writing update for %s: %s", key, err)
 			}
-			p.notifyWatchers(key, msg.GetData())
+			p.notifyWatchers(key, data)
 		}
 	}
 }
 
+func (p *PubsubValueStore) handleNewMsgs(ctx context.Context, sub *pubsub.Subscription, key string) ([]byte, error) {
+	msg, err := sub.Next(ctx)
+	if err != nil {
+		if err != context.Canceled {
+			log.Warningf("PubsubResolve: subscription error in %s: %s", key, err.Error())
+		}
+		return nil, err
+	}
+	return msg.GetData(), nil
+}
+
+func (p *PubsubValueStore) handleNewPeer(ctx context.Context, sub *pubsub.Subscription, key string) ([]byte, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	var pid peer.ID
+
+	for {
+		peerEvt, err := sub.NextPeerEvent(ctx)
+		if err != nil {
+			if err != context.Canceled {
+				log.Warningf("PubsubNewPeer: subscription error in %s: %s", key, err.Error())
+			}
+			return nil, err
+		}
+		if peerEvt.Type == pubsub.PeerJoin {
+			pid = peerEvt.Peer
+			break
+		}
+	}
+
+	return p.fetch.Get(ctx, pid, key)
```

Comment:
In the [spec](https://github.com/ipfs/specs/pull/218#discussion_r314271523) corresponding to this PR @lidel mentioned he thought that sending out requests immediately on PeerJoin could lead to a node getting overwhelmed by connections when they first Subscribe to a topic.

While I think there are some optimizations that could be done here. However, I'm not convinced they're really necessary given that if the data being fetched is relatively small (true for IPNS records) it shouldn't add much more burden then the PubSub stream negotiation + initial Subscribe messages.

@Stebalien @vyzo Thoughts?

Reply:
It is a concern, but I don't know how much of a problem will be in reality.

Reply:
That's my though as well, we can always come back and add more complexity later.